### PR TITLE
#87 continuous integration

### PR DESCRIPTION
**Pull request by WalrusArtist**
The travis.yml file is required in order to accompany continuous integration using travisCI

### CODE REVIEW

**Comment by platisd** [comment]
I'd say add everything in the same pull request and let me know when it's ready. Feel free to ask for help or feedback when you need it.

**Comment by WalrusArtist** [comment]
I see. I was wondering why I'm unable to see that my repository is running after I have added the yml file on travis. It is stated that I have no builds for the repository.

**Comment by platisd** [comment]
Rename the configuration file to `.travis.yml`. You've got an extra "a" in the file name. :+1:

**Comment by WalrusArtist** [comment]
@platisd I have updated my yml file but travis seems to only detect my previous commit rather than my last one. How do I make it so that it detects my last commit?

File: arduino/obstacleAvoidance/obstacleAvoidance.ino [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // very basic obstacle avoidance
 // stop the car when an obstacle gets to close to the front sensor
 
-#include <Smartcar.h>
```
Comment (platisd) [inline arduino/obstacleAvoidance/obstacleAvoidance.ino]: I don't think removing this library will help you compile this sketch! :grin:

**Comment by platisd** [comment]
Remove that `.yaml` file completely and see if anything changes

**Comment by WalrusArtist** [comment]
@platisd I have been trying to test my ladder commits and had found out that travis couldn't parse any of them. I looked around and found out that it could be due to indentation problems (using tabs over spaces). I had converted to spaces using Atom and Sublime with no luck. I am unable to detect my fault at this point.

Screenshot 2019-08-30 at 14 42 35

**Comment by platisd** [comment]
Can you send me a link to the actual logs of the failed build because I can't find them?

**Comment by platisd** [comment]
@WalrusArtist the commit I just pushed fixes any formatting errors (actually I didn't understand why Travis was unhappy with what you had but anyway) among some other smaller mistakes. See the commit's description and contents for more details.

In any case, now it builds the Python and the Arduino layers of the project. This means that the first requirement of the re-exam is fulfilled. There's still an error with the app, which I am have no idea how to solve since i'm clueless about anything iOS. If it gives you too much trouble just remove building the application and focus on deploying the Arduino app as a release.

**Comment by WalrusArtist** [comment]
> deploying the Arduino app as a release.

@platisd Would TestFlight be enough for this? If you're not familiar with what it is, it's basically an air-to-air deployment service for client testings provided by Apple.

**Comment by platisd** [comment]
> > deploying the Arduino app as a release.
> 
> @platisd Would TestFlight be enough for this? If you're not familiar with what it is, it's basically an air-to-air deployment service for client testings provided by Apple.

Hmmm, I'm not sure how TestFlight works. I'd like a compiled version of one component of the system (aside the python script) to become available automatically every time something gets merged to `master`.
If this TestFlight can do that it's good with me.

**Comment by platisd** [comment]
The easiest way as far as I'm concerned is to do that via the Github releases. In the issue (#87) there's a tutorial on how to achieve that. Any other way that works is good enough for me.

**Comment by WalrusArtist** [comment]
@platisd To obtain the hex file I would need to compile the .ino file. And for some reason i'm getting an error that states that there's no such file or directory "Smartcar.h". I'm unaware of why this is.

**Comment by WalrusArtist** [comment]
Never mind, I just found out for some reason the library wasn't installed anymore

> @platisd To obtain the hex file I would need to compile the .ino file. And for some reason i'm getting an error that states that there's no such file or directory "Smartcar.h". I'm unaware of why this is.

**Comment by platisd** [comment]
Not sure what you saw but since the build didn't fail the sketches were built.
Now you'll have to figure out where the compiled sketch ends up, point the deployed `file` field there and you'll be done I believe.

**Comment by WalrusArtist** [comment]
@platisd I was not able to obtain the hex file using Serial1 from obstacleAvoidance, and Serial3 from testCar. Changed both to Serial and they both compiled successfully resulting along the hex file. I hope this is okay

**Comment by platisd** [comment]
No, its not OK to make changes in the logic so to make the build pass. :stuck_out_tongue:
What is the error exactly?

**Comment by WalrusArtist** [comment]
Fair enough :) the error seems to be a basic declaration one. I'm not sure if it's due to a library I am missing, or maybe my Arduino version? Attached is a screenshot 

Screenshot 2019-08-31 at 21 58 29

**Comment by platisd** [comment]
OK I see. You're trying to compile this on your machine. Well, you have to choose "Arduino Mega" as your board. On CI it builds correctly since the `mega2560` argument is passed to `build_platform`.

The directories you're trying to find the hex files seem "temporary". How did you find out about them?

File: arduino/testCar/testCar.ino [hunk 2]
```diff
@@ -15,7 +15,7 @@ SR04 frontUltrasound(US_TRIGGER_PIN, US_ECHO_PIN, US_MAX_DISTANCE);   // ultra s
 
 // HeadingCar and PathFinder derived from it
 HeadingCar car(control, gyroscope);
-PathFinder pathy(car, &Serial3, &leftOdometer, &rightOdometer, &frontUltrasound, Point(30, 2));
+PathFinder pathy(car, &Serial, &leftOdometer, &rightOdometer, &frontUltrasound, Point(30, 2));
```
Comment (platisd) [inline arduino/testCar/testCar.ino]: Don't forget this one

**Review by platisd — CHANGES_REQUESTED** [review]

**Comment by WalrusArtist** [comment]
> The directories you're trying to find the hex files seem "temporary". How did you find out about them?

@platisd In the Arduino IDE selecting Sketch > Export Compiled Binary, and looking at the third line of the transactions in the compiler was the only the hex file i could find. 

I changed it to mega and it compiled successfully, and I got the hex files. Are you saying these hex files won't do because they're also temporary?

**Comment by WalrusArtist** [comment]
> > The directories you're trying to find the hex files seem "temporary". How did you find out about them?
> 
> @platisd In the Arduino IDE selecting Sketch > Export Compiled Binary, and looking at the third line of the transactions in the compiler was the only the hex file i could find.
> 
> I changed it to mega and it compiled successfully, and I got the hex files. Are you saying these hex files won't do because they're also temporary?

FYI, the paths were the same so i didn't need to reedit the yaml file

**Comment by platisd** [comment]
OK, then these file paths are most likely wrong. I I suspect that the hex files might be in the same folder. Try printing looking for them with something like `find . -iname "*.hex* 2>/dev/null` after you run the build step to see if you can locate them in your workspace.

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+          - ~/arduino_ide
+          - ~/.arduino15/packages/
+
+deploy:
+  provider: releases
+  api_key:
+    secure: bdeac7bd7ffa0b46a10afbc198cdd6254d2cc6d1
+  file:
+    - /arduino/testCar/testCar.ino.with_bootloader.mega.hex
+    - /arduino/testCar/testCar.ino.mega.hex
```
Comment (platisd) [inline .travis.yml]: These two paths look wrong. You should probably add `$TRAVIS_BUILD_DIR` in front of them.

**Review by platisd — CHANGES_REQUESTED** [review]

**Comment by platisd** [comment]
Yeah, apparently the [CI script from Adafruit](https://github.com/adafruit/travis-ci-arduino) does not preserve the `.hex` or any other temporary files created during compilation. I made a [modified version](https://github.com/platisd/travis-ci-arduino) of it (which I will try to get merged with the Adafruit one) which does keep them.

The configuration now should be ready to be merged.
Please squash the commits before merging them so you don't flood the git history! (i.e. don't click on "Merge pull request" but on the arrow next to it and choose the appropriate option)
To test everything out after you have merged this, [create a new release](https://github.com/DIT112-V19/group-04/releases/new).

**Review by platisd — CHANGES_REQUESTED** [review]
Last thing you need to do:
* Remove `app/CarInterface.xcodeproj/project.xcworkspace/xcuserdata/kardodastin.xcuserdatad/UserInterfaceState.xcuserstate`

**Review by platisd — APPROVED** [review]
Good job

**Pull request merged**